tests/test_single_topology_v3.py

```diff
@@ -383,3 +390,113 @@ def test_no_chiral_bond_restraints():
     assert len(state.chiral_bond.get_idxs()) == 0
     U = state.get_U_fn()
     _ = U(init_conf)
+
+
+def assert_interpolation_valid(f, a, b):
+    np.testing.assert_allclose(f(a, b, 0.0), a)
+    np.testing.assert_allclose(f(a, b, 1.0), b)
+
+    rng = np.random.default_rng(2022)
+    lambdas = rng.uniform(0, 1, (30,))
+    ys = vmap(functools.partial(f, a, b))(lambdas)
+    assert np.all(np.minimum(a, b) <= ys)
+    assert np.all(ys <= np.maximum(a, b))
+
+
+def test_handle_ring_opening_closing():
+    rng = np.random.default_rng(2022)
+    n = 100
+    k = 20
+
+    def random_ks():
+        return rng.uniform(1, 2, (n,))
+
+    src_k = random_ks()
+    dst_k = random_ks()
+
+    idxs = rng.choice(n, k, replace=False)
+
+    closing_idxs = idxs[: k // 2]
+    opening_idxs = idxs[k // 2 :]
+
+    src_k[closing_idxs] = 0.0
+    dst_k[opening_idxs] = 0.0
+
+    lambda_min = 0.3
+    lambda_max = 0.7
+
+    f = functools.partial(
+        handle_ring_opening_closing,
+        linear_interpolation,
+        src_k,
+        dst_k,
+        lambda_min=lambda_min,
+        lambda_max=lambda_max,
+    )
+
+    # closing
+    # 0 < λ < λmin
+    ks = f(0.5 * lambda_min)
+    np.testing.assert_array_equal(ks[closing_idxs], 0.0)
+
+    # λmax < λ < 1
+    ks = f(0.5 * (1.0 + lambda_max))
+    np.testing.assert_array_equal(ks[closing_idxs], dst_k[closing_idxs])
+
+    # opening
+    # 1 - λmin < λ < 1
+    ks = f(0.5 * (2.0 - lambda_min))
+    np.testing.assert_array_equal(ks[opening_idxs], 0.0)
+
+    # 0 < λ < 1 - λmax
+    ks = f(0.5 * (1.0 - lambda_max))
+    np.testing.assert_array_equal(ks[opening_idxs], src_k[opening_idxs])
+
+
+def test_interpolate_harmonic_force_constant():
+
+    rng = np.random.default_rng(2022)
+
+    def random_ks():
+        return rng.uniform(1, 2, (100,))
+
+    src_k = random_ks()
+    dst_k = random_ks()
+
+    k_min = 0.1
+    assert (k_min < np.minimum(src_k, dst_k)).all()
+
+    f = functools.partial(
+        interpolate_harmonic_force_constant,
+        k_min=k_min,
+        lambda_min=0.0,
+        lambda_max=0.4,
+    )
+
+    assert_interpolation_valid(f, src_k, dst_k)
+
+    # check for sublinearity
+    lam = 0.1
+    k_linear = linear_interpolation(src_k, dst_k, lam)
+    k_loglinear = f(src_k, dst_k, lam)
+    assert np.all(k_loglinear < k_linear)
+
+    # check for JIT-compatibility
+    f = jax.jit(f)
+    _ = f(src_k, dst_k, 0.1)
+
+
+def test_cyclic_difference():
+    assert cyclic_difference(0, 0, 1) == 0
+    assert cyclic_difference(0, 1, 2) == 1  # arbitrary, positive by convention
+    assert cyclic_difference(0, 0, 3) == 0
+    assert cyclic_difference(0, 1, 3) == 1
+    assert cyclic_difference(0, 2, 3) == -1
+    _ = jax.jit(cyclic_difference)(0, 1, 1)
```

Comment:
should we add a test 

1) with some negative numbers?  I'm not sure if we're guaranteed positive phases in the OFF spec
2) handle the case where accidentally b < a?

Reply:
Should be addressed by [7960a0a](https://github.com/proteneer/timemachine/pull/820/commits/7960a0ad1e7b977e8760a3ed5fd682b9004b3bac)